File: oop-encapsulation/test/service/StudentService.test.ts
import { StudentService } from '../../src/service/StudentService';
import { StudentRepository } from '../../src/repository/StudentRepository';

describe('StudentService', () => {
  let studentService: StudentService;
  let studentRepository: StudentRepository;

  beforeEach(() => {
    studentRepository = {
      save: jest.fn(),
    };
    studentService = new StudentService(studentRepository);
  });

  test('should save student with free tuition', async () => {
    const id = 1;
    const name = 'Any name';
    const score = 97.0;

    await studentService.create(id, name, score);

    expect(studentRepository.save).toHaveBeenCalledWith(
      expect.objectContaining({
        id: id,
        name: name,
        score: score,
        isEligibleForFreeTuition: true,
      }),
    );
  });

  test('should save student without free tuition', async () => {
    const id = 1;
    const name = 'Any name';
    const score = 70;

    await studentService.create(id, name, score);

    expect(studentRepository.save).toHaveBeenCalledWith(
      expect.objectContaining({
        id: id,
        name: name,
        score: score,
        isEligibleForFreeTuition: false,
      }),
    );
  });

  test('should not save student with negative score', async () => {
    const id = 1;
    const name = 'Any name';
    const score = -50;

    await expect(async () => {
      await studentService.create(id, name, score);
    }).rejects.toThrow('Score must be between 0 and 100');
  });

  test('should not save student with invalid score', async () => {
    const id = 1;
    const name = 'Any name';
    const score = 200;

    await expect(async () => {
      await studentService.create(id, name, score);
    }).rejects.toThrow('Score must be between 0 and 100');
  });
});

File: oop-encapsulation/ruby/test/service/student_service_test.rb
require 'rspec/autorun'
require_relative '../../src/service/student_service'
require_relative '../../src/repository/student_repository'

describe StudentService do
  let(:studentRepository) { double('StudentRepository') }
  let(:studentService) { StudentService.new(studentRepository) }

  it 'should save student with free tuition' do
    id = 1
    name = 'Any name'
    score = 97.0

    expect(studentRepository).to receive(:save).with(
      an_object_having_attributes(
        id: id,
        name: name,
        score: score,
        is_eligible_for_free_tuition: true
      )
    )

    studentService.create(id, name, score)
  end

  it 'should save student without free tuition' do
    id = 1
    name = 'Any name'
    score = 70

    expect(studentRepository).to receive(:save).with(
      an_object_having_attributes(
        id: id,
        name: name,
        score: score,
        is_eligible_for_free_tuition: false
      )
    )

    studentService.create(id, name, score)
  end

  it 'should not save student with negative score' do
    id = 1
    name = 'Any name'
    score = -50

    expect {
      studentService.create(id, name, score)
    }.to raise_error('Score must be between 0 and 100')
  end

  it 'should not save student with invalid score' do
    id = 1
    name = 'Any name'
    score = 200

    expect {
      studentService.create(id, name, score)
    }.to raise_error('Score must be between 0 and 100')
  end
end
